### PR TITLE
fix: handle localhost and port in asWebviewUri

### LINKED ISSUE #130151: asWebviewUri doesn't return the correct Uri when remote authority is localhost:[port]
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: 1.58.2
- OS Version: macOS 11.4

Steps to Reproduce:

I wrote a test case inside VS Code which reproduces the error. Then I added a patch which fixes it. 

This is what `asWebviewUri` currently returns (bug) `https://vscode-remote%2Blocalhost:8080.vscode-resource.vscode-webview.net/Users/codey/file.html` and then is what it should return with the fix `https://vscode-remote%2Blocalhost.vscode-resource.vscode-webview.net:8080/Users/codey/file.html`

### PR DESCRIPTION
**Pull request by jsjoeio**
This PR fixes #130151

We encountered this error in code-server (fix here: https://github.com/cdr/code-server/pull/3895) and noticed that resources weren't loading in webviews due to the wrong URI being returned from `asWebviewUri`. 

I don't think Codespaces has the issue because it's always running on a standard port and so it won't have a port in the url.

## Additional Context

We noticed this happening with a custom VS Code extension having issues loading resources (CSS & JS) in webviews.

I was going to try and reproduce with `vscode` by running `yarn web` but I'm not seeing the option to manually install VSIX files 🤔

image

### CODE REVIEW

**Comment by jsjoeio** [comment]
Not sure how to fix these 😅
```logs
[build/lib/layersChecker.ts]: Reference to 'URL' from 'lib.dom.d.ts' violates layer '**/vs/**/common/**' (/home/runner/work/vscode/vscode/src/vs/workbench/api/common/shared/webview.ts (60,28)
[build/lib/layersChecker.ts]: Reference to 'hostname' from 'lib.dom.d.ts' violates layer '**/vs/**/common/**' (/home/runner/work/vscode/vscode/src/vs/workbench/api/common/shared/webview.ts (61,50)
[build/lib/layersChecker.ts]: Reference to 'port' from 'lib.dom.d.ts' violates layer '**/vs/**/common/**' (/home/runner/work/vscode/vscode/src/vs/workbench/api/common/shared/webview.ts (61,106)
[build/lib/layersChecker.ts]: Reference to 'port' from 'lib.dom.d.ts' violates layer '**/vs/**/common/**' (/home/runner/work/vscode/vscode/src/vs/workbench/api/common/shared/webview.ts (61,133)
```

File: src/vs/workbench/api/common/shared/webview.ts [hunk 1]
```diff
@@ -55,9 +55,15 @@ export function asWebviewUri(
 		});
 	}
 
+	let authority = `${resource.scheme}+${resource.authority}.${webviewRootResourceAuthority}`;
+	if (resource.authority.includes('localhost')) {
+		const authorityUrl = new URL(`http://${resource.authority}`);
```
Comment (mjbvz) [inline src/vs/workbench/api/common/shared/webview.ts:60]: To fix the layer errors, try using our `URI` class instead of the built-in `URL` class

Comment (jsjoeio) [reply src/vs/workbench/api/common/shared/webview.ts:60]: I think I tried that before but the `URI` class doesn't have a port (though it does have it in the `authority` but then I'd have to split it on the `:`, right?) Unless I leave the port out entirely?

Comment (Emyrk) [reply src/vs/workbench/api/common/shared/webview.ts:60]: @mjbvz The URI puts the `authority` section before the `vscode-resource.vscode-webview.net`. So the url becomes `localhost:8080.vscode-resource.vscode-webview.net` which is an invalid port.

Ideally, the `URI` package could parse the port like `URL`.

**Comment by jsjoeio** [comment]
@mjbvz is this patch/fix no longer needed thanks to https://github.com/microsoft/vscode/commit/649dd18019332e4dc9090abcf0b98bccf8d57b85?

**Comment by mjbvz** [comment]
Possibly? Although I didn't test ports specifically with that fix

Did you have a chance to test the issue from #130151?

**Comment by jsjoeio** [comment]
I spoke to @code-asher who tested it and I believe it could fix it:

```
> encodeAuthority("localhost:8080")
"localhost-003a8080"

> decodeAuthority("localhost-003a8080")
"localhost:8080"

> new URL("http://localhost-003a8080")
URL { href: "http://localhost-003a8080/", origin: "http://localhost-003a8080", protocol: "http:", username: "", password: "", host: "localhost-003a8080", hostname: "localhost-003a8080", port: "", pathname: "/", search: "" }
```

I will have to double-check and follow-up. If it does, I could instead add new tests for URLs with ports

**Comment by jsjoeio** [comment]
Closing in favor of https://github.com/microsoft/vscode/pull/131447

**Pull request closed without merging**